### PR TITLE
CORE-8875: Mutual TLS allowed client certificates subject API

### PR DESCRIPTION
**Pull request by yift-r3**
Adds API to maintain the MGM-allowed certificates subject list. Those APIs should work only for an MGM node and only when the Gateway is configured to support mutual TLS. The subjects are kept in the MGM virtual node database.

The API pull request can be found at https://github.com/corda/corda-api/pull/773

### CODE REVIEW

File: components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt [hunk 1]
```diff
@@ -42,4 +44,57 @@ interface MGMRpcOps : RpcOps {
         @HttpRpcPathParameter(description = "The holding identity ID of the MGM of the membership group to be joined")
         holdingIdentityShortHash: String
     ): String
+
+    /**
+     * Adds client certificate subject to the mutual TLS allow client certificates.
```
Comment (williamvigorr3) [inline components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt]: Change `allow client certificates` too `allowed client certificates` to be the same as `mutualTlsDisallowClientCertificate` (I think this is clearer)?

File: components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt [hunk 2]
```diff
@@ -42,4 +44,57 @@ interface MGMRpcOps : RpcOps {
...
+    )
+
+    /**
+     * Remove client certificate subject from the mutual TLS group allowed client certificates.
+     *
+     * @param holdingIdentityShortHash The holding identity ID of the MGM.
+     * @param subject The certificate subject.
+     */
+    @HttpRpcDELETE(
+        path = "{holdingIdentityShortHash}/mutual-tls/allow-client/{subject}",
```
Comment (williamvigorr3) [inline components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt]: I would change `allow-client` in the path to  `allowed-client-certificate-subjects`. I think that would make these requests clearer, even if it makes everything more verbose..

File: components/membership/membership-http-rpc-impl/src/test/kotlin/net/corda/membership/impl/httprpc/v1/MGMRpcOpsTest.kt [hunk 3]
```diff
@@ -46,10 +51,19 @@ class MGMRpcOpsTest {
     private val mgmOpsClient: MGMOpsClient = mock {
         on { generateGroupPolicy(any()) } doReturn mgmGenerateGroupPolicyResponseDto
     }
+    private val subject = "CN=Alice, O=Alice ,L=London ,C=GB"
+    private val gatewayConfiguration = mock<SmartConfig> {
+        on { getConfig("sslConfig") } doReturn mock
+        on { getString("tlsType") } doReturn "MUTUAL"
```
Comment (williamvigorr3) [inline components/membership/membership-http-rpc-impl/src/test/kotlin/net/corda/membership/impl/httprpc/v1/MGMRpcOpsTest.kt:57]: Is this definitely correct? I think `tlsType` should be under the `sslConfig` and not the top level gateway config?

File: components/membership/membership-persistence-client-impl/src/main/kotlin/net/corda/membership/impl/persistence/client/MembershipQueryClientImpl.kt [hunk 4]
```diff
@@ -216,4 +218,23 @@ class MembershipQueryClientImpl(
...
+            MutualTlsListAllowedCertificates()
+        ).execute()
+        return when (val payload = result.payload) {
+            is MutualTlsListAllowedCertificatesResponse -> {
+                MembershipQueryResult.Success(
+                    payload.subjects,
+                )
+            }
+            else -> {
+                MembershipQueryResult.Failure("Failed to get retrieve list of allowed certificates.")
```
Comment (williamvigorr3) [inline components/membership/membership-persistence-client-impl/src/main/kotlin/net/corda/membership/impl/persistence/client/MembershipQueryClientImpl.kt]: `Failed to retrieve list of allowed certificates.`?

**Review by williamvigorr3 — COMMENTED** [review]
I left some minor comments

File: components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt (hunk 2 above)
Comment (yift-r3) [reply components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt]: @corda/http-rpc - any preferences?

Comment (vkolomeyko) [reply components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt]: Either `allow-client` or `allowed-client-certificate-subjects` is fine from my point of view.

File: components/membership/membership-http-rpc-impl/src/test/kotlin/net/corda/membership/impl/httprpc/v1/MGMRpcOpsTest.kt (hunk 3 above)
Comment (yift-r3) [reply components/membership/membership-http-rpc-impl/src/test/kotlin/net/corda/membership/impl/httprpc/v1/MGMRpcOpsTest.kt:57]: The line above it (`on { getConfig("sslConfig") } doReturn mock`) makes the same mock the mock for the gateway and the tls...

Comment (williamvigorr3) [reply components/membership/membership-http-rpc-impl/src/test/kotlin/net/corda/membership/impl/httprpc/v1/MGMRpcOpsTest.kt:57]: Shouldn't we test with the correct structure? They are not supposed to be the same?

Comment (yift-r3) [reply components/membership/membership-http-rpc-impl/src/test/kotlin/net/corda/membership/impl/httprpc/v1/MGMRpcOpsTest.kt:57]: We have a proper test for this in `getClusterType returns the correct value if available`

File: components/membership/membership-persistence-client-impl/src/main/kotlin/net/corda/membership/impl/persistence/client/MembershipQueryClientImpl.kt (hunk 4 above)
Comment (williamvigorr3) [reply components/membership/membership-persistence-client-impl/src/main/kotlin/net/corda/membership/impl/persistence/client/MembershipQueryClientImpl.kt]: This is unaddressed.

**Review by charlieR3 — DISMISSED** [review]
Has some build issues but code looks good to me otherwise

**Review by charlieR3 — DISMISSED** [review]

**Review by williamvigorr3 — DISMISSED** [review]
LGTM

File: components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt [hunk 5]
```diff
@@ -42,4 +44,57 @@ interface MGMRpcOps : RpcOps {
...
+    )
+
+    /**
+     * List the allowed client certificate subjects for mutual TLS.
+     *
+     * @param holdingIdentityShortHash The holding identity ID of the MGM.
+     * @return List of the allowed client certificate subjects.
+     */
+    @HttpRpcGET(
+        path = "{holdingIdentityShortHash}/mutual-tls/allowed-client-certificate-subjects/",
```
Comment (vkolomeyko) [inline components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/MGMRpcOps.kt]: Minor:
```suggestion
        path = "{holdingIdentityShortHash}/mutual-tls/allowed-client-certificate-subjects",
```

**Review by ronanbrowne — APPROVED** [review]
LGTM - build

**Review by williamvigorr3 — APPROVED** [review]
LGTM

**Review by vkolomeyko — APPROVED** [review]
LGTM

**Pull request merged**